// @sentry/vite-plugin/dist/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var bundlerPluginCore = require('@sentry/bundler-plugin-core');

function viteReleaseInjectionPlugin(injectionCode) {
  return {
    name: "sentry-vite-release-injection-plugin",
    enforce: "pre",
    // need this so that vite runs the resolveId hook
    vite: bundlerPluginCore.createRollupReleaseInjectionHooks(injectionCode)
  };
}
function viteDebugIdInjectionPlugin() {
  return {
    name: "sentry-vite-debug-id-injection-plugin",
    vite: bundlerPluginCore.createRollupDebugIdInjectionHooks()
  };
}
function viteDebugIdUploadPlugin(upload) {
  return {
    name: "sentry-vite-debug-id-upload-plugin",
    vite: bundlerPluginCore.createRollupDebugIdUploadHooks(upload)
  };
}
var sentryUnplugin = bundlerPluginCore.sentryUnpluginFactory({
  releaseInjectionPlugin: viteReleaseInjectionPlugin,
  debugIdInjectionPlugin: viteDebugIdInjectionPlugin,
  debugIdUploadPlugin: viteDebugIdUploadPlugin
});

// eslint-disable-next-line @typescript-eslint/no-explicit-any
var sentryVitePlugin = sentryUnplugin.vite;

Object.defineProperty(exports, 'sentryCliBinaryExists', {
  enumerable: true,
  get: function () { return bundlerPluginCore.sentryCliBinaryExists; }
});
exports.sentryVitePlugin = sentryVitePlugin;
