// posthog-js/lib/src/extensions/sentry-integration.js
/**
 * Integrate Sentry with PostHog. This will add a direct link to the person in Sentry, and an $exception event in PostHog
 *
 * ### Usage
 *
 *     Sentry.init({
 *          dsn: 'https://example',
 *          integrations: [
 *              new posthog.SentryIntegration(posthog)
 *          ]
 *     })
 *
 * @param {Object} [posthog] The posthog object
 * @param {string} [organization] Optional: The Sentry organization, used to send a direct link from PostHog to Sentry
 * @param {Number} [projectId] Optional: The Sentry project id, used to send a direct link from PostHog to Sentry
 * @param {string} [prefix] Optional: Url of a self-hosted sentry instance (default: https://sentry.io/organizations/)
 */
var SentryIntegration = /** @class */ (function () {
    function SentryIntegration(_posthog, organization, projectId, prefix) {
        // setupOnce gets called by Sentry when it intializes the plugin
        this.name = 'posthog-js';
        this.setupOnce = function (addGlobalEventProcessor) {
            addGlobalEventProcessor(function (event) {
                var _a, _b, _c, _d, _e;
                if (event.level !== 'error' || !_posthog.__loaded)
                    return event;
                if (!event.tags)
                    event.tags = {};
                var host = _posthog.config.ui_host || _posthog.config.api_host;
                event.tags['PostHog Person URL'] = host + '/person/' + _posthog.get_distinct_id();
                if (_posthog.sessionRecordingStarted()) {
                    event.tags['PostHog Recording URL'] = _posthog.get_session_replay_url({ withTimestamp: true });
                }
                var exceptions = ((_a = event.exception) === null || _a === void 0 ? void 0 : _a.values) || [];
                var data = {
                    // PostHog Exception Properties,
                    $exception_message: (_b = exceptions[0]) === null || _b === void 0 ? void 0 : _b.value,
                    $exception_type: (_c = exceptions[0]) === null || _c === void 0 ? void 0 : _c.type,
                    $exception_personURL: host + '/person/' + _posthog.get_distinct_id(),
                    // Sentry Exception Properties
                    $sentry_event_id: event.event_id,
                    $sentry_exception: event.exception,
                    $sentry_exception_message: (_d = exceptions[0]) === null || _d === void 0 ? void 0 : _d.value,
                    $sentry_exception_type: (_e = exceptions[0]) === null || _e === void 0 ? void 0 : _e.type,
                    $sentry_tags: event.tags,
                };
                if (organization && projectId)
                    data['$sentry_url'] =
                        (prefix || 'https://sentry.io/organizations/') +
                            organization +
                            '/issues/?project=' +
                            projectId +
                            '&query=' +
                            event.event_id;
                _posthog.capture('$exception', data);
                return event;
            });
        };
    }
    return SentryIntegration;
}());
export { SentryIntegration };
