// posthog-js/lib/src/extensions/exception-autocapture/type-checking.js
import { _isFunction, _isNull, _isObject, _isUndefined } from '../../utils/type-utils';
export function isEvent(candidate) {
    return !_isUndefined(Event) && isInstanceOf(candidate, Event);
}
export function isPlainObject(candidate) {
    return isBuiltin(candidate, 'Object');
}
export function isInstanceOf(candidate, base) {
    try {
        return candidate instanceof base;
    }
    catch (_e) {
        return false;
    }
}
export function isPrimitive(candidate) {
    return _isNull(candidate) || (!_isObject(candidate) && !_isFunction(candidate));
}
export function isError(candidate) {
    switch (Object.prototype.toString.call(candidate)) {
        case '[object Error]':
        case '[object Exception]':
        case '[object DOMException]':
            return true;
        default:
            return isInstanceOf(candidate, Error);
    }
}
export function isErrorEvent(event) {
    return isBuiltin(event, 'ErrorEvent');
}
export function isErrorWithStack(candidate) {
    return 'stack' in candidate;
}
export function isBuiltin(candidate, className) {
    return Object.prototype.toString.call(candidate) === "[object ".concat(className, "]");
}
export function isDOMException(candidate) {
    return isBuiltin(candidate, 'DOMException');
}
export function isDOMError(candidate) {
    return isBuiltin(candidate, 'DOMError');
}
