// @sentry/integrations/cjs/index.js
Object.defineProperty(exports, '__esModule', { value: true });

const captureconsole = require('./captureconsole.js');
const debug = require('./debug.js');
const dedupe = require('./dedupe.js');
const extraerrordata = require('./extraerrordata.js');
const offline = require('./offline.js');
const reportingobserver = require('./reportingobserver.js');
const rewriteframes = require('./rewriteframes.js');
const sessiontiming = require('./sessiontiming.js');
const transaction = require('./transaction.js');
const httpclient = require('./httpclient.js');



exports.CaptureConsole = captureconsole.CaptureConsole;
exports.Debug = debug.Debug;
exports.Dedupe = dedupe.Dedupe;
exports.ExtraErrorData = extraerrordata.ExtraErrorData;
exports.Offline = offline.Offline;
exports.ReportingObserver = reportingobserver.ReportingObserver;
exports.RewriteFrames = rewriteframes.RewriteFrames;
exports.SessionTiming = sessiontiming.SessionTiming;
exports.Transaction = transaction.Transaction;
exports.HttpClient = httpclient.HttpClient;
