// posthog-js/lib/src/constants.js
/*
 * Constants
 */
/* PROPERTY KEYS */
// This key is deprecated, but we want to check for it to see whether aliasing is allowed.
export var PEOPLE_DISTINCT_ID_KEY = '$people_distinct_id';
export var DISTINCT_ID = 'distinct_id';
export var ALIAS_ID_KEY = '__alias';
export var CAMPAIGN_IDS_KEY = '__cmpns';
export var EVENT_TIMERS_KEY = '__timers';
export var AUTOCAPTURE_DISABLED_SERVER_SIDE = '$autocapture_disabled_server_side';
export var SESSION_RECORDING_ENABLED_SERVER_SIDE = '$session_recording_enabled_server_side';
export var CONSOLE_LOG_RECORDING_ENABLED_SERVER_SIDE = '$console_log_recording_enabled_server_side';
export var SESSION_RECORDING_RECORDER_VERSION_SERVER_SIDE = '$session_recording_recorder_version_server_side'; // follows rrweb versioning
export var SESSION_RECORDING_NETWORK_PAYLOAD_CAPTURE = '$session_recording_network_payload_capture';
export var SESSION_ID = '$sesid';
export var SESSION_RECORDING_IS_SAMPLED = '$session_is_sampled';
export var ENABLED_FEATURE_FLAGS = '$enabled_feature_flags';
export var PERSISTENCE_EARLY_ACCESS_FEATURES = '$early_access_features';
export var STORED_PERSON_PROPERTIES_KEY = '$stored_person_properties';
export var STORED_GROUP_PROPERTIES_KEY = '$stored_group_properties';
export var SURVEYS = '$surveys';
export var FLAG_CALL_REPORTED = '$flag_call_reported';
export var USER_STATE = '$user_state';
export var POSTHOG_QUOTA_LIMITED = '$posthog_quota_limited';
export var CLIENT_SESSION_PROPS = '$client_session_props';
// These are properties that are reserved and will not be automatically included in events
export var PERSISTENCE_RESERVED_PROPERTIES = [
    PEOPLE_DISTINCT_ID_KEY,
    ALIAS_ID_KEY,
    CAMPAIGN_IDS_KEY,
    EVENT_TIMERS_KEY,
    SESSION_RECORDING_ENABLED_SERVER_SIDE,
    SESSION_ID,
    ENABLED_FEATURE_FLAGS,
    USER_STATE,
    POSTHOG_QUOTA_LIMITED,
    PERSISTENCE_EARLY_ACCESS_FEATURES,
    STORED_GROUP_PROPERTIES_KEY,
    STORED_PERSON_PROPERTIES_KEY,
    SURVEYS,
    FLAG_CALL_REPORTED,
    CLIENT_SESSION_PROPS,
];
