// posthog-js/lib/src/page-view.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { window } from './utils/globals';
var PageViewManager = /** @class */ (function () {
    function PageViewManager() {
        var _this = this;
        this._hasSeenPageView = false;
        this._updateScrollData = function () {
            var _a, _b, _c, _d;
            if (!_this._pageViewData) {
                _this._pageViewData = _this._createPageViewData();
            }
            var pageViewData = _this._pageViewData;
            var scrollY = _this._scrollY();
            var scrollHeight = _this._scrollHeight();
            var contentY = _this._contentY();
            var contentHeight = _this._contentHeight();
            pageViewData.lastScrollY = scrollY;
            pageViewData.maxScrollY = Math.max(scrollY, (_a = pageViewData.maxScrollY) !== null && _a !== void 0 ? _a : 0);
            pageViewData.maxScrollHeight = Math.max(scrollHeight, (_b = pageViewData.maxScrollHeight) !== null && _b !== void 0 ? _b : 0);
            pageViewData.lastContentY = contentY;
            pageViewData.maxContentY = Math.max(contentY, (_c = pageViewData.maxContentY) !== null && _c !== void 0 ? _c : 0);
            pageViewData.maxContentHeight = Math.max(contentHeight, (_d = pageViewData.maxContentHeight) !== null && _d !== void 0 ? _d : 0);
        };
    }
    PageViewManager.prototype._createPageViewData = function () {
        var _a;
        return {
            pathname: (_a = window === null || window === void 0 ? void 0 : window.location.pathname) !== null && _a !== void 0 ? _a : '',
        };
    };
    PageViewManager.prototype.doPageView = function () {
        var prevPageViewData;
        // if there were events created before the first PageView, we would have created a
        // pageViewData for them. If this happened, we don't want to create a new pageViewData
        if (!this._hasSeenPageView) {
            this._hasSeenPageView = true;
            prevPageViewData = undefined;
            if (!this._pageViewData) {
                this._pageViewData = this._createPageViewData();
            }
        }
        else {
            prevPageViewData = this._pageViewData;
            this._pageViewData = this._createPageViewData();
        }
        // update the scroll properties for the new page, but wait until the next tick
        // of the event loop
        setTimeout(this._updateScrollData, 0);
        return __assign({ $prev_pageview_pathname: prevPageViewData === null || prevPageViewData === void 0 ? void 0 : prevPageViewData.pathname }, this._calculatePrevPageScrollProperties(prevPageViewData));
    };
    PageViewManager.prototype.doPageLeave = function () {
        var prevPageViewData = this._pageViewData;
        return __assign({ $prev_pageview_pathname: prevPageViewData === null || prevPageViewData === void 0 ? void 0 : prevPageViewData.pathname }, this._calculatePrevPageScrollProperties(prevPageViewData));
    };
    PageViewManager.prototype._calculatePrevPageScrollProperties = function (prevPageViewData) {
        if (!prevPageViewData ||
            prevPageViewData.maxScrollHeight == null ||
            prevPageViewData.lastScrollY == null ||
            prevPageViewData.maxScrollY == null ||
            prevPageViewData.maxContentHeight == null ||
            prevPageViewData.lastContentY == null ||
            prevPageViewData.maxContentY == null) {
            return {};
        }
        var maxScrollHeight = prevPageViewData.maxScrollHeight, lastScrollY = prevPageViewData.lastScrollY, maxScrollY = prevPageViewData.maxScrollY, maxContentHeight = prevPageViewData.maxContentHeight, lastContentY = prevPageViewData.lastContentY, maxContentY = prevPageViewData.maxContentY;
        // Use ceil, so that e.g. scrolling 999.5px of a 1000px page is considered 100% scrolled
        maxScrollHeight = Math.ceil(maxScrollHeight);
        lastScrollY = Math.ceil(lastScrollY);
        maxScrollY = Math.ceil(maxScrollY);
        maxContentHeight = Math.ceil(maxContentHeight);
        lastContentY = Math.ceil(lastContentY);
        maxContentY = Math.ceil(maxContentY);
        // if the maximum scroll height is near 0, then the percentage is 1
        var lastScrollPercentage = maxScrollHeight <= 1 ? 1 : clamp(lastScrollY / maxScrollHeight, 0, 1);
        var maxScrollPercentage = maxScrollHeight <= 1 ? 1 : clamp(maxScrollY / maxScrollHeight, 0, 1);
        var lastContentPercentage = maxContentHeight <= 1 ? 1 : clamp(lastContentY / maxContentHeight, 0, 1);
        var maxContentPercentage = maxContentHeight <= 1 ? 1 : clamp(maxContentY / maxContentHeight, 0, 1);
        return {
            $prev_pageview_last_scroll: lastScrollY,
            $prev_pageview_last_scroll_percentage: lastScrollPercentage,
            $prev_pageview_max_scroll: maxScrollY,
            $prev_pageview_max_scroll_percentage: maxScrollPercentage,
            $prev_pageview_last_content: lastContentY,
            $prev_pageview_last_content_percentage: lastContentPercentage,
            $prev_pageview_max_content: maxContentY,
            $prev_pageview_max_content_percentage: maxContentPercentage,
        };
    };
    PageViewManager.prototype.startMeasuringScrollPosition = function () {
        window === null || window === void 0 ? void 0 : window.addEventListener('scroll', this._updateScrollData);
        window === null || window === void 0 ? void 0 : window.addEventListener('scrollend', this._updateScrollData);
        window === null || window === void 0 ? void 0 : window.addEventListener('resize', this._updateScrollData);
    };
    PageViewManager.prototype.stopMeasuringScrollPosition = function () {
        window === null || window === void 0 ? void 0 : window.removeEventListener('scroll', this._updateScrollData);
        window === null || window === void 0 ? void 0 : window.removeEventListener('scrollend', this._updateScrollData);
        window === null || window === void 0 ? void 0 : window.removeEventListener('resize', this._updateScrollData);
    };
    PageViewManager.prototype._scrollHeight = function () {
        return window
            ? Math.max(0, window.document.documentElement.scrollHeight - window.document.documentElement.clientHeight)
            : 0;
    };
    PageViewManager.prototype._scrollY = function () {
        return window ? window.scrollY || window.pageYOffset || window.document.documentElement.scrollTop || 0 : 0;
    };
    PageViewManager.prototype._contentHeight = function () {
        return (window === null || window === void 0 ? void 0 : window.document.documentElement.scrollHeight) || 0;
    };
    PageViewManager.prototype._contentY = function () {
        var clientHeight = (window === null || window === void 0 ? void 0 : window.document.documentElement.clientHeight) || 0;
        return this._scrollY() + clientHeight;
    };
    return PageViewManager;
}());
export { PageViewManager };
function clamp(x, min, max) {
    return Math.max(min, Math.min(x, max));
}
