// posthog-js/lib/src/send-request.js
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
import { _each } from './utils';
import Config from './config';
import { _HTTPBuildQuery } from './utils/request-utils';
import { _isArray, _isFunction, _isNumber, _isUint8Array, _isUndefined } from './utils/type-utils';
import { logger } from './utils/logger';
export var addParamsToURL = function (url, urlQueryArgs, parameterOptions) {
    var e_1, _a;
    var args = urlQueryArgs || {};
    args['ip'] = parameterOptions['ip'] ? 1 : 0;
    args['_'] = new Date().getTime().toString();
    args['ver'] = Config.LIB_VERSION;
    var halves = url.split('?');
    if (halves.length > 1) {
        var params = halves[1].split('&');
        try {
            for (var params_1 = __values(params), params_1_1 = params_1.next(); !params_1_1.done; params_1_1 = params_1.next()) {
                var p = params_1_1.value;
                var key = p.split('=')[0];
                if (!_isUndefined(args[key])) {
                    delete args[key];
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (params_1_1 && !params_1_1.done && (_a = params_1.return)) _a.call(params_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    var argSeparator = url.indexOf('?') > -1 ? '&' : '?';
    return url + argSeparator + _HTTPBuildQuery(args);
};
export var encodePostData = function (data, options) {
    if (options.blob && data.buffer) {
        return new Blob([_isUint8Array(data) ? data : data.buffer], { type: 'text/plain' });
    }
    if (options.sendBeacon || options.blob) {
        var body = encodePostData(data, { method: 'POST' });
        return new Blob([body], { type: 'application/x-www-form-urlencoded' });
    }
    if (options.method !== 'POST') {
        return null;
    }
    var body_data;
    if (_isArray(data) || _isUint8Array(data)) {
        // TODO: eh? passing an Array here?
        body_data = 'data=' + encodeURIComponent(data);
    }
    else {
        body_data = 'data=' + encodeURIComponent(data.data);
    }
    if ('compression' in data && data.compression) {
        body_data += '&compression=' + data.compression;
    }
    return body_data;
};
export var xhr = function (_a) {
    var url = _a.url, data = _a.data, headers = _a.headers, options = _a.options, callback = _a.callback, retriesPerformedSoFar = _a.retriesPerformedSoFar, retryQueue = _a.retryQueue, onXHRError = _a.onXHRError, _b = _a.timeout, timeout = _b === void 0 ? 60000 : _b, onResponse = _a.onResponse;
    if (_isNumber(retriesPerformedSoFar) && retriesPerformedSoFar > 0) {
        url = addParamsToURL(url, { retry_count: retriesPerformedSoFar }, {});
    }
    var req = new XMLHttpRequest();
    req.open(options.method || 'GET', url, true);
    var body = encodePostData(data, options);
    _each(headers, function (headerValue, headerName) {
        req.setRequestHeader(headerName, headerValue);
    });
    if (options.method === 'POST' && !options.blob) {
        req.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
    }
    req.timeout = timeout;
    // send the ph_optout cookie
    // withCredentials cannot be modified until after calling .open on Android and Mobile Safari
    req.withCredentials = true;
    req.onreadystatechange = function () {
        // XMLHttpRequest.DONE == 4, except in safari 4
        if (req.readyState === 4) {
            onResponse === null || onResponse === void 0 ? void 0 : onResponse(req);
            if (req.status === 200) {
                if (callback) {
                    var response = void 0;
                    try {
                        response = JSON.parse(req.responseText);
                    }
                    catch (e) {
                        logger.error(e);
                        return;
                    }
                    callback(response);
                }
            }
            else {
                if (_isFunction(onXHRError)) {
                    onXHRError(req);
                }
                // don't retry errors between 400 and 500 inclusive
                if (req.status < 400 || req.status > 500) {
                    retryQueue.enqueue({
                        url: url,
                        data: data,
                        options: options,
                        headers: headers,
                        retriesPerformedSoFar: (retriesPerformedSoFar || 0) + 1,
                        callback: callback,
                    });
                }
                callback === null || callback === void 0 ? void 0 : callback({ status: 0 });
            }
        }
    };
    req.send(body);
};
