// posthog-js/lib/src/session-props.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
/* Client-side session parameters. These are primarily used by web analytics,
 * which relies on these for session analytics without the plugin server being
 * available for the person level set-once properties. Obviously not consistent
 * between client-side events and server-side events but this is acceptable
 * as web analytics only uses client-side.
 *
 * These have the same lifespan as a session_id
 */
import { window } from './utils/globals';
import { _info } from './utils/event-utils';
import { CLIENT_SESSION_PROPS } from './constants';
var generateSessionSourceParams = function () {
    return __assign({ initialPathName: (window === null || window === void 0 ? void 0 : window.location.pathname) || '', referringDomain: _info.referringDomain() }, _info.campaignParams());
};
var SessionPropsManager = /** @class */ (function () {
    function SessionPropsManager(sessionIdManager, persistence, sessionSourceParamGenerator) {
        var _this = this;
        this._onSessionIdCallback = function (sessionId) {
            var _a;
            var stored = _this._getStoredProps();
            if (stored && stored.sessionId === sessionId) {
                return;
            }
            var newProps = {
                sessionId: sessionId,
                props: _this._sessionSourceParamGenerator(),
            };
            _this._persistence.register((_a = {}, _a[CLIENT_SESSION_PROPS] = newProps, _a));
        };
        this._sessionIdManager = sessionIdManager;
        this._persistence = persistence;
        this._sessionSourceParamGenerator = sessionSourceParamGenerator || generateSessionSourceParams;
        this._sessionIdManager.onSessionId(this._onSessionIdCallback);
    }
    SessionPropsManager.prototype._getStoredProps = function () {
        return this._persistence.props[CLIENT_SESSION_PROPS];
    };
    SessionPropsManager.prototype.getSessionProps = function () {
        var _a;
        var p = (_a = this._getStoredProps()) === null || _a === void 0 ? void 0 : _a.props;
        if (!p) {
            return {};
        }
        return {
            $client_session_initial_referring_host: p.referringDomain,
            $client_session_initial_pathname: p.initialPathName,
            $client_session_initial_utm_source: p.utm_source,
            $client_session_initial_utm_campaign: p.utm_campaign,
            $client_session_initial_utm_medium: p.utm_medium,
            $client_session_initial_utm_content: p.utm_content,
            $client_session_initial_utm_term: p.utm_term,
        };
    };
    return SessionPropsManager;
}());
export { SessionPropsManager };
