// posthog-js/lib/src/extensions/replay/config.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { _isFunction } from '../../utils/type-utils';
import { convertToURL } from '../../utils/request-utils';
import { logger } from '../../utils/logger';
export var defaultNetworkOptions = {
    initiatorTypes: [
        'audio',
        'beacon',
        'body',
        'css',
        'early-hint',
        'embed',
        'fetch',
        'frame',
        'iframe',
        'icon',
        'image',
        'img',
        'input',
        'link',
        'navigation',
        'object',
        'ping',
        'script',
        'track',
        'video',
        'xmlhttprequest',
    ],
    maskRequestFn: function (data) { return data; },
    recordHeaders: false,
    recordBody: false,
    recordInitialRequests: false,
    recordPerformance: false,
    performanceEntryTypeToObserve: [
        // 'event', // This is too noisy as it covers all browser events
        'first-input',
        // 'mark', // Mark is used too liberally. We would need to filter for specific marks
        // 'measure', // Measure is used too liberally. We would need to filter for specific measures
        'navigation',
        'paint',
        'resource',
    ],
    payloadSizeLimitBytes: 1000000,
};
var HEADER_DENYLIST = [
    'authorization',
    'x-forwarded-for',
    'authorization',
    'cookie',
    'set-cookie',
    'x-api-key',
    'x-real-ip',
    'remote-addr',
    'forwarded',
    'proxy-authorization',
    'x-csrf-token',
    'x-csrftoken',
    'x-xsrf-token',
];
// we always remove headers on the deny list because we never want to capture this sensitive data
var removeAuthorizationHeader = function (data) {
    var _a;
    Object.keys((_a = data.requestHeaders) !== null && _a !== void 0 ? _a : {}).forEach(function (header) {
        var _a;
        if (HEADER_DENYLIST.includes(header.toLowerCase()))
            (_a = data.requestHeaders) === null || _a === void 0 ? true : delete _a[header];
    });
    return data;
};
var POSTHOG_PATHS_TO_IGNORE = ['/s/', '/e/', '/i/vo/e/'];
// want to ignore posthog paths when capturing requests, or we can get trapped in a loop
// because calls to PostHog would be reported using a call to PostHog which would be reported....
var ignorePostHogPaths = function (data) {
    var url = convertToURL(data.name);
    if (url && url.pathname && POSTHOG_PATHS_TO_IGNORE.includes(url.pathname)) {
        return undefined;
    }
    return data;
};
function redactPayload(payload, headers, limit, description) {
    var requestContentLength = headers === null || headers === void 0 ? void 0 : headers['content-length'];
    // in the interests of bundle size and the complexity of estimating payload size
    // we only check the content-length header if it's present
    // this might mean we can't always limit the payload, but that's better than
    // having lots of code shipped to every browser that will rarely run
    if (requestContentLength && parseInt(requestContentLength) > limit) {
        return "".concat(description, " body too large to record");
    }
    return payload;
}
// people can have arbitrarily large payloads on their site, but we don't want to ingest them
var limitPayloadSize = function (options) {
    var _a;
    // the smallest of 1MB or the specified limit if there is one
    var limit = Math.min(1000000, (_a = options.payloadSizeLimitBytes) !== null && _a !== void 0 ? _a : 1000000);
    return function (data) {
        if (data === null || data === void 0 ? void 0 : data.requestBody) {
            data.requestBody = redactPayload(data.requestBody, data.requestHeaders, limit, 'Request');
        }
        if (data === null || data === void 0 ? void 0 : data.responseBody) {
            data.responseBody = redactPayload(data.responseBody, data.responseHeaders, limit, 'Response');
        }
        return data;
    };
};
/**
 *  whether a maskRequestFn is provided or not,
 *  we ensure that we remove the denied header from requests
 *  we _never_ want to record that header by accident
 *  if someone complains then we'll add an opt-in to let them override it
 */
export var buildNetworkRequestOptions = function (instanceConfig, remoteNetworkOptions) {
    var config = instanceConfig.session_recording;
    // client can always disable despite remote options
    var canRecordHeaders = config.recordHeaders === false ? false : remoteNetworkOptions.recordHeaders;
    var canRecordBody = config.recordBody === false ? false : remoteNetworkOptions.recordBody;
    var canRecordPerformance = config.recordPerformance === false ? false : remoteNetworkOptions.recordPerformance;
    var payloadLimiter = limitPayloadSize(config);
    var enforcedCleaningFn = function (d) {
        return payloadLimiter(ignorePostHogPaths(removeAuthorizationHeader(d)));
    };
    var hasDeprecatedMaskFunction = _isFunction(instanceConfig.session_recording.maskNetworkRequestFn);
    if (hasDeprecatedMaskFunction && _isFunction(instanceConfig.session_recording.maskCapturedNetworkRequestFn)) {
        logger.warn('Both `maskNetworkRequestFn` and `maskCapturedNetworkRequestFn` are defined. `maskNetworkRequestFn` will be ignored.');
    }
    if (hasDeprecatedMaskFunction) {
        instanceConfig.session_recording.maskCapturedNetworkRequestFn = function (data) {
            var cleanedURL = instanceConfig.session_recording.maskNetworkRequestFn({ url: data.name });
            return __assign(__assign({}, data), { name: cleanedURL === null || cleanedURL === void 0 ? void 0 : cleanedURL.url });
        };
    }
    config.maskRequestFn = _isFunction(instanceConfig.session_recording.maskCapturedNetworkRequestFn)
        ? function (data) {
            var _a, _b, _c;
            var cleanedRequest = enforcedCleaningFn(data);
            return cleanedRequest
                ? (_c = (_b = (_a = instanceConfig.session_recording).maskCapturedNetworkRequestFn) === null || _b === void 0 ? void 0 : _b.call(_a, cleanedRequest)) !== null && _c !== void 0 ? _c : undefined
                : undefined;
        }
        : undefined;
    if (!config.maskRequestFn) {
        config.maskRequestFn = enforcedCleaningFn;
    }
    return __assign(__assign(__assign({}, defaultNetworkOptions), config), { recordHeaders: canRecordHeaders, recordBody: canRecordBody, recordPerformance: canRecordPerformance, recordInitialRequests: canRecordPerformance });
};
