// posthog-js/lib/src/extensions/replay/sessionrecording.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
import { CONSOLE_LOG_RECORDING_ENABLED_SERVER_SIDE, SESSION_RECORDING_ENABLED_SERVER_SIDE, SESSION_RECORDING_IS_SAMPLED, SESSION_RECORDING_NETWORK_PAYLOAD_CAPTURE, SESSION_RECORDING_RECORDER_VERSION_SERVER_SIDE, } from '../../constants';
import { FULL_SNAPSHOT_EVENT_TYPE, INCREMENTAL_SNAPSHOT_EVENT_TYPE, META_EVENT_TYPE, MutationRateLimiter, truncateLargeConsoleLogs, } from './sessionrecording-utils';
import { EventType } from '@rrweb/types';
import Config from '../../config';
import { _timestamp, loadScript } from '../../utils';
import { _isBoolean, _isFunction, _isNull, _isNumber, _isObject, _isString, _isUndefined } from '../../utils/type-utils';
import { logger } from '../../utils/logger';
import { assignableWindow, window } from '../../utils/globals';
import { buildNetworkRequestOptions } from './config';
import { isLocalhost } from '../../utils/request-utils';
var BASE_ENDPOINT = '/s/';
export var RECORDING_IDLE_ACTIVITY_TIMEOUT_MS = 5 * 60 * 1000; // 5 minutes
export var RECORDING_MAX_EVENT_SIZE = 1024 * 1024 * 0.9; // ~1mb (with some wiggle room)
export var RECORDING_BUFFER_TIMEOUT = 2000; // 2 seconds
export var SESSION_RECORDING_BATCH_KEY = 'recordings';
// NOTE: Importing this type is problematic as we can't safely bundle it to a TS definition so, instead we redefine.
// import type { record } from 'rrweb2/typings'
// import type { recordOptions } from 'rrweb/typings/types'
// Copied from rrweb typings to avoid import
var IncrementalSource;
(function (IncrementalSource) {
    IncrementalSource[IncrementalSource["Mutation"] = 0] = "Mutation";
    IncrementalSource[IncrementalSource["MouseMove"] = 1] = "MouseMove";
    IncrementalSource[IncrementalSource["MouseInteraction"] = 2] = "MouseInteraction";
    IncrementalSource[IncrementalSource["Scroll"] = 3] = "Scroll";
    IncrementalSource[IncrementalSource["ViewportResize"] = 4] = "ViewportResize";
    IncrementalSource[IncrementalSource["Input"] = 5] = "Input";
    IncrementalSource[IncrementalSource["TouchMove"] = 6] = "TouchMove";
    IncrementalSource[IncrementalSource["MediaInteraction"] = 7] = "MediaInteraction";
    IncrementalSource[IncrementalSource["StyleSheetRule"] = 8] = "StyleSheetRule";
    IncrementalSource[IncrementalSource["CanvasMutation"] = 9] = "CanvasMutation";
    IncrementalSource[IncrementalSource["Font"] = 10] = "Font";
    IncrementalSource[IncrementalSource["Log"] = 11] = "Log";
    IncrementalSource[IncrementalSource["Drag"] = 12] = "Drag";
    IncrementalSource[IncrementalSource["StyleDeclaration"] = 13] = "StyleDeclaration";
    IncrementalSource[IncrementalSource["Selection"] = 14] = "Selection";
    IncrementalSource[IncrementalSource["AdoptedStyleSheet"] = 15] = "AdoptedStyleSheet";
})(IncrementalSource || (IncrementalSource = {}));
var ACTIVE_SOURCES = [
    IncrementalSource.MouseMove,
    IncrementalSource.MouseInteraction,
    IncrementalSource.Scroll,
    IncrementalSource.ViewportResize,
    IncrementalSource.Input,
    IncrementalSource.TouchMove,
    IncrementalSource.MediaInteraction,
    IncrementalSource.Drag,
];
var SessionRecording = /** @class */ (function () {
    function SessionRecording(instance) {
        var _this = this;
        this.isIdle = false;
        this._linkedFlagSeen = false;
        this._lastActivityTimestamp = Date.now();
        this.windowId = null;
        this.sessionId = null;
        this._linkedFlag = null;
        this._sampleRate = null;
        this._minimumDuration = null;
        // Util to help developers working on this feature manually override
        this._forceAllowLocalhostNetworkCapture = false;
        this.instance = instance;
        this._captureStarted = false;
        this._endpoint = BASE_ENDPOINT;
        this.stopRrweb = undefined;
        this.receivedDecide = false;
        window === null || window === void 0 ? void 0 : window.addEventListener('beforeunload', function () {
            _this._flushBuffer();
        });
        if (!this.instance.sessionManager) {
            logger.error('Session recording started without valid sessionManager');
            throw new Error('Session recording started without valid sessionManager. This is a bug.');
        }
        this.buffer = this.clearBuffer();
    }
    Object.defineProperty(SessionRecording.prototype, "started", {
        get: function () {
            // TODO could we use status instead of _captureStarted?
            return this._captureStarted;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SessionRecording.prototype, "sessionManager", {
        get: function () {
            if (!this.instance.sessionManager) {
                logger.error('Session recording started without valid sessionManager');
                throw new Error('Session recording started without valid sessionManager. This is a bug.');
            }
            return this.instance.sessionManager;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SessionRecording.prototype, "isSampled", {
        get: function () {
            if (_isNumber(this._sampleRate)) {
                return this.instance.get_property(SESSION_RECORDING_IS_SAMPLED);
            }
            else {
                return null;
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SessionRecording.prototype, "sessionDuration", {
        get: function () {
            var _a, _b;
            var mostRecentSnapshot = (_a = this.buffer) === null || _a === void 0 ? void 0 : _a.data[((_b = this.buffer) === null || _b === void 0 ? void 0 : _b.data.length) - 1];
            var sessionStartTimestamp = this.sessionManager.checkAndGetSessionAndWindowId(true).sessionStartTimestamp;
            return mostRecentSnapshot ? mostRecentSnapshot.timestamp - sessionStartTimestamp : null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SessionRecording.prototype, "isRecordingEnabled", {
        get: function () {
            var enabled_server_side = !!this.instance.get_property(SESSION_RECORDING_ENABLED_SERVER_SIDE);
            var enabled_client_side = !this.instance.config.disable_session_recording;
            return window && enabled_server_side && enabled_client_side;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SessionRecording.prototype, "isConsoleLogCaptureEnabled", {
        get: function () {
            var enabled_server_side = !!this.instance.get_property(CONSOLE_LOG_RECORDING_ENABLED_SERVER_SIDE);
            var enabled_client_side = this.instance.config.enable_recording_console_log;
            return enabled_client_side !== null && enabled_client_side !== void 0 ? enabled_client_side : enabled_server_side;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SessionRecording.prototype, "recordingVersion", {
        get: function () {
            var _a;
            var recordingVersion_server_side = this.instance.get_property(SESSION_RECORDING_RECORDER_VERSION_SERVER_SIDE);
            var recordingVersion_client_side = (_a = this.instance.config.session_recording) === null || _a === void 0 ? void 0 : _a.recorderVersion;
            return recordingVersion_client_side || recordingVersion_server_side || 'v1';
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SessionRecording.prototype, "networkPayloadCapture", {
        // network payload capture config has three parts
        // each can be configured server side or client side
        get: function () {
            var _a, _b;
            var networkPayloadCapture_server_side = this.instance.get_property(SESSION_RECORDING_NETWORK_PAYLOAD_CAPTURE);
            var networkPayloadCapture_client_side = {
                recordHeaders: (_a = this.instance.config.session_recording) === null || _a === void 0 ? void 0 : _a.recordHeaders,
                recordBody: (_b = this.instance.config.session_recording) === null || _b === void 0 ? void 0 : _b.recordBody,
            };
            var headersEnabled = (networkPayloadCapture_client_side === null || networkPayloadCapture_client_side === void 0 ? void 0 : networkPayloadCapture_client_side.recordHeaders) || (networkPayloadCapture_server_side === null || networkPayloadCapture_server_side === void 0 ? void 0 : networkPayloadCapture_server_side.recordHeaders);
            var bodyEnabled = (networkPayloadCapture_client_side === null || networkPayloadCapture_client_side === void 0 ? void 0 : networkPayloadCapture_client_side.recordBody) || (networkPayloadCapture_server_side === null || networkPayloadCapture_server_side === void 0 ? void 0 : networkPayloadCapture_server_side.recordBody);
            var performanceEnabled = this.instance.config.capture_performance || (networkPayloadCapture_server_side === null || networkPayloadCapture_server_side === void 0 ? void 0 : networkPayloadCapture_server_side.capturePerformance);
            return headersEnabled || bodyEnabled || performanceEnabled
                ? { recordHeaders: headersEnabled, recordBody: bodyEnabled, recordPerformance: performanceEnabled }
                : undefined;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(SessionRecording.prototype, "status", {
        /**
         * defaults to buffering mode until a decide response is received
         * once a decide response is received status can be disabled, active or sampled
         */
        get: function () {
            if (!this.receivedDecide) {
                return 'buffering';
            }
            if (!this.isRecordingEnabled) {
                return 'disabled';
            }
            if (_isString(this._linkedFlag) && !this._linkedFlagSeen) {
                return 'buffering';
            }
            if (_isBoolean(this.isSampled)) {
                return this.isSampled ? 'sampled' : 'disabled';
            }
            else {
                return 'active';
            }
        },
        enumerable: false,
        configurable: true
    });
    SessionRecording.prototype.startRecordingIfEnabled = function () {
        if (this.isRecordingEnabled) {
            this.startCaptureAndTrySendingQueuedSnapshots();
        }
        else {
            this.stopRecording();
            this.clearBuffer();
        }
    };
    SessionRecording.prototype.stopRecording = function () {
        if (this._captureStarted && this.stopRrweb) {
            this.stopRrweb();
            this.stopRrweb = undefined;
            this._captureStarted = false;
        }
    };
    SessionRecording.prototype.makeSamplingDecision = function (sessionId) {
        var _a, _b;
        var _c, _d;
        var sessionIdChanged = this.sessionId !== sessionId;
        if (!_isNumber(this._sampleRate)) {
            (_c = this.instance.persistence) === null || _c === void 0 ? void 0 : _c.register((_a = {},
                _a[SESSION_RECORDING_IS_SAMPLED] = null,
                _a));
            return;
        }
        var storedIsSampled = this.isSampled;
        /**
         * if we get this far then we should make a sampling decision.
         * When the session id changes or there is no stored sampling decision for this session id
         * then we should make a new decision.
         *
         * Otherwise, we should use the stored decision.
         */
        var shouldSample;
        if (sessionIdChanged || !_isBoolean(storedIsSampled)) {
            var randomNumber = Math.random();
            shouldSample = randomNumber < this._sampleRate;
        }
        else {
            shouldSample = storedIsSampled;
        }
        if (!shouldSample) {
            logger.warn("[SessionSampling] Sample rate (".concat(this._sampleRate, ") has determined that this sessionId (").concat(sessionId, ") will not be sent to the server."));
        }
        (_d = this.instance.persistence) === null || _d === void 0 ? void 0 : _d.register((_b = {},
            _b[SESSION_RECORDING_IS_SAMPLED] = shouldSample,
            _b));
    };
    SessionRecording.prototype.afterDecideResponse = function (response) {
        var _a;
        var _this = this;
        var _b, _c, _d, _e, _f, _g, _h, _j;
        if (this.instance.persistence) {
            this.instance.persistence.register((_a = {},
                _a[SESSION_RECORDING_ENABLED_SERVER_SIDE] = !!response['sessionRecording'],
                _a[CONSOLE_LOG_RECORDING_ENABLED_SERVER_SIDE] = (_b = response.sessionRecording) === null || _b === void 0 ? void 0 : _b.consoleLogRecordingEnabled,
                _a[SESSION_RECORDING_RECORDER_VERSION_SERVER_SIDE] = (_c = response.sessionRecording) === null || _c === void 0 ? void 0 : _c.recorderVersion,
                _a[SESSION_RECORDING_NETWORK_PAYLOAD_CAPTURE] = __assign({ capturePerformance: response.capturePerformance }, (_d = response.sessionRecording) === null || _d === void 0 ? void 0 : _d.networkPayloadCapture),
                _a));
        }
        var receivedSampleRate = (_e = response.sessionRecording) === null || _e === void 0 ? void 0 : _e.sampleRate;
        this._sampleRate =
            _isUndefined(receivedSampleRate) || _isNull(receivedSampleRate) ? null : parseFloat(receivedSampleRate);
        var receivedMinimumDuration = (_f = response.sessionRecording) === null || _f === void 0 ? void 0 : _f.minimumDurationMilliseconds;
        this._minimumDuration = _isUndefined(receivedMinimumDuration) ? null : receivedMinimumDuration;
        this._linkedFlag = ((_g = response.sessionRecording) === null || _g === void 0 ? void 0 : _g.linkedFlag) || null;
        if ((_h = response.sessionRecording) === null || _h === void 0 ? void 0 : _h.endpoint) {
            this._endpoint = (_j = response.sessionRecording) === null || _j === void 0 ? void 0 : _j.endpoint;
        }
        if (_isNumber(this._sampleRate)) {
            this.sessionManager.onSessionId(function (sessionId) {
                _this.makeSamplingDecision(sessionId);
            });
        }
        if (_isString(this._linkedFlag)) {
            var linkedFlag_1 = this._linkedFlag;
            this.instance.onFeatureFlags(function (flags) {
                _this._linkedFlagSeen = flags.includes(linkedFlag_1);
            });
        }
        this.receivedDecide = true;
        this.startRecordingIfEnabled();
    };
    SessionRecording.prototype.log = function (message, level) {
        var _a;
        if (level === void 0) { level = 'log'; }
        (_a = this.instance.sessionRecording) === null || _a === void 0 ? void 0 : _a.onRRwebEmit({
            type: 6,
            data: {
                plugin: 'rrweb/console@1',
                payload: {
                    level: level,
                    trace: [],
                    // Even though it is a string we stringify it as that's what rrweb expects
                    payload: [JSON.stringify(message)],
                },
            },
            timestamp: _timestamp(),
        });
    };
    SessionRecording.prototype.startCaptureAndTrySendingQueuedSnapshots = function () {
        this._startCapture();
    };
    SessionRecording.prototype._startCapture = function () {
        var _this = this;
        if (_isUndefined(Object.assign)) {
            // According to the rrweb docs, rrweb is not supported on IE11 and below:
            // "rrweb does not support IE11 and below because it uses the MutationObserver API which was supported by these browsers."
            // https://github.com/rrweb-io/rrweb/blob/master/guide.md#compatibility-note
            //
            // However, MutationObserver does exist on IE11, it just doesn't work well and does not detect all changes.
            // Instead, when we load "recorder.js", the first JS error is about "Object.assign" being undefined.
            // Thus instead of MutationObserver, we look for this function and block recording if it's undefined.
            return;
        }
        // We do not switch recorder versions midway through a recording.
        if (this._captureStarted || this.instance.config.disable_session_recording) {
            return;
        }
        this._captureStarted = true;
        // We want to ensure the sessionManager is reset if necessary on load of the recorder
        this.sessionManager.checkAndGetSessionAndWindowId();
        var recorderJS = this.recordingVersion === 'v2' ? 'recorder-v2.js' : 'recorder.js';
        // If recorder.js is already loaded (if array.full.js snippet is used or posthog-js/dist/recorder is
        // imported) or matches the requested recorder version, don't load script. Otherwise, remotely import
        // recorder.js from cdn since it hasn't been loaded.
        if (this.instance.__loaded_recorder_version !== this.recordingVersion) {
            loadScript(this.instance.config.api_host + "/static/".concat(recorderJS, "?v=").concat(Config.LIB_VERSION), function (err) {
                if (err) {
                    return logger.error("Could not load ".concat(recorderJS), err);
                }
                _this._onScriptLoaded();
            });
        }
        else {
            this._onScriptLoaded();
        }
    };
    SessionRecording.prototype._isInteractiveEvent = function (event) {
        var _a;
        return event.type === INCREMENTAL_SNAPSHOT_EVENT_TYPE && ACTIVE_SOURCES.indexOf((_a = event.data) === null || _a === void 0 ? void 0 : _a.source) !== -1;
    };
    SessionRecording.prototype._updateWindowAndSessionIds = function (event) {
        // Some recording events are triggered by non-user events (e.g. "X minutes ago" text updating on the screen).
        // We don't want to extend the session or trigger a new session in these cases. These events are designated by event
        // type -> incremental update, and source -> mutation.
        var isUserInteraction = this._isInteractiveEvent(event);
        if (!isUserInteraction && !this.isIdle) {
            // We check if the lastActivityTimestamp is old enough to go idle
            if (event.timestamp - this._lastActivityTimestamp > RECORDING_IDLE_ACTIVITY_TIMEOUT_MS) {
                this.isIdle = true;
                this._tryAddCustomEvent('sessionIdle', {
                    reason: 'user inactivity',
                    timeSinceLastActive: event.timestamp - this._lastActivityTimestamp,
                    threshold: RECORDING_IDLE_ACTIVITY_TIMEOUT_MS,
                });
            }
        }
        if (isUserInteraction) {
            this._lastActivityTimestamp = event.timestamp;
            if (this.isIdle) {
                // Remove the idle state if set and trigger a full snapshot as we will have ignored previous mutations
                this.isIdle = false;
                this._tryAddCustomEvent('sessionNoLongerIdle', {
                    reason: 'user activity',
                    type: event.type,
                });
                this._tryTakeFullSnapshot();
            }
        }
        if (this.isIdle) {
            return;
        }
        // We only want to extend the session if it is an interactive event.
        var _a = this.sessionManager.checkAndGetSessionAndWindowId(!isUserInteraction, event.timestamp), windowId = _a.windowId, sessionId = _a.sessionId;
        var sessionIdChanged = this.sessionId !== sessionId;
        var windowIdChanged = this.windowId !== windowId;
        if ([FULL_SNAPSHOT_EVENT_TYPE, META_EVENT_TYPE].indexOf(event.type) === -1 &&
            (windowIdChanged || sessionIdChanged)) {
            this._tryTakeFullSnapshot();
        }
        this.windowId = windowId;
        this.sessionId = sessionId;
    };
    SessionRecording.prototype._tryRRwebMethod = function (rrwebMethod) {
        if (!this._captureStarted) {
            return false;
        }
        try {
            rrwebMethod();
            return true;
        }
        catch (e) {
            // Sometimes a race can occur where the recorder is not fully started yet
            logger.error('[Session-Recording] using rrweb when not started.', e);
            return false;
        }
    };
    SessionRecording.prototype._tryAddCustomEvent = function (tag, payload) {
        var _this = this;
        return this._tryRRwebMethod(function () { var _a; return (_a = _this.rrwebRecord) === null || _a === void 0 ? void 0 : _a.addCustomEvent(tag, payload); });
    };
    SessionRecording.prototype._tryTakeFullSnapshot = function () {
        var _this = this;
        return this._tryRRwebMethod(function () { var _a; return (_a = _this.rrwebRecord) === null || _a === void 0 ? void 0 : _a.takeFullSnapshot(); });
    };
    SessionRecording.prototype._onScriptLoaded = function () {
        var e_1, _a;
        var _this = this;
        var _b;
        // rrweb config info: https://github.com/rrweb-io/rrweb/blob/7d5d0033258d6c29599fb08412202d9a2c7b9413/src/record/index.ts#L28
        var sessionRecordingOptions = {
            // select set of rrweb config options we expose to our users
            // see https://github.com/rrweb-io/rrweb/blob/master/guide.md
            blockClass: 'ph-no-capture',
            blockSelector: undefined,
            ignoreClass: 'ph-ignore-input',
            maskTextClass: 'ph-mask',
            maskTextSelector: undefined,
            maskTextFn: undefined,
            maskAllInputs: true,
            maskInputOptions: {},
            maskInputFn: undefined,
            slimDOMOptions: {},
            collectFonts: false,
            inlineStylesheet: true,
            recordCrossOriginIframes: false,
        };
        // We switched from loading all of rrweb to just the record part, but
        // keep backwards compatibility if someone hasn't upgraded PostHog
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        this.rrwebRecord = window.rrweb ? window.rrweb.record : window.rrwebRecord;
        // only allows user to set our allow-listed options
        var userSessionRecordingOptions = this.instance.config.session_recording;
        try {
            for (var _c = __values(Object.entries(userSessionRecordingOptions || {})), _d = _c.next(); !_d.done; _d = _c.next()) {
                var _e = __read(_d.value, 2), key = _e[0], value = _e[1];
                if (key in sessionRecordingOptions) {
                    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                    // @ts-ignore
                    sessionRecordingOptions[key] = value;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (!this.rrwebRecord) {
            logger.error('onScriptLoaded was called but rrwebRecord is not available. This indicates something has gone wrong.');
            return;
        }
        this.mutationRateLimiter =
            (_b = this.mutationRateLimiter) !== null && _b !== void 0 ? _b : new MutationRateLimiter(this.rrwebRecord, {
                onBlockedNode: function (id, node) {
                    var message = "Too many mutations on node '".concat(id, "'. Rate limiting. This could be due to SVG animations or something similar");
                    logger.info(message, {
                        node: node,
                    });
                    _this.log('[PostHog Recorder] ' + message, 'warn');
                },
            });
        this.stopRrweb = this.rrwebRecord(__assign({ emit: function (event) {
                _this.onRRwebEmit(event);
            }, plugins: this._gatherRRWebPlugins() }, sessionRecordingOptions));
        // :TRICKY: rrweb does not capture navigation within SPA-s, so hook into our $pageview events to get access to all events.
        //   Dropping the initial event is fine (it's always captured by rrweb).
        this.instance._addCaptureHook(function (eventName) {
            // If anything could go wrong here it has the potential to block the main loop,
            // so we catch all errors.
            try {
                if (eventName === '$pageview') {
                    var href = window ? _this._maskUrl(window.location.href) : '';
                    if (!href) {
                        return;
                    }
                    _this._tryAddCustomEvent('$pageview', { href: href });
                }
            }
            catch (e) {
                logger.error('Could not add $pageview to rrweb session', e);
            }
        });
        // We reset the last activity timestamp, resetting the idle timer
        this._lastActivityTimestamp = Date.now();
        this.isIdle = false;
    };
    SessionRecording.prototype._gatherRRWebPlugins = function () {
        var plugins = [];
        if (assignableWindow.rrwebConsoleRecord && this.isConsoleLogCaptureEnabled) {
            plugins.push(assignableWindow.rrwebConsoleRecord.getRecordConsolePlugin());
        }
        if (this.networkPayloadCapture && _isFunction(assignableWindow.getRecordNetworkPlugin)) {
            var canRecordNetwork = !isLocalhost() || this._forceAllowLocalhostNetworkCapture;
            if (canRecordNetwork) {
                plugins.push(assignableWindow.getRecordNetworkPlugin(buildNetworkRequestOptions(this.instance.config, this.networkPayloadCapture)));
            }
            else {
                logger.info('[SessionReplay-NetworkCapture] not started because we are on localhost.');
            }
        }
        return plugins;
    };
    SessionRecording.prototype.onRRwebEmit = function (rawEvent) {
        if (!rawEvent || !_isObject(rawEvent)) {
            return;
        }
        if (rawEvent.type === EventType.Meta) {
            var href = this._maskUrl(rawEvent.data.href);
            if (!href) {
                return;
            }
            rawEvent.data.href = href;
        }
        var throttledEvent = this.mutationRateLimiter
            ? this.mutationRateLimiter.throttleMutations(rawEvent)
            : rawEvent;
        if (!throttledEvent) {
            return;
        }
        // TODO: Re-add ensureMaxMessageSize once we are confident in it
        var event = truncateLargeConsoleLogs(throttledEvent);
        var size = JSON.stringify(event).length;
        this._updateWindowAndSessionIds(event);
        if (this.isIdle) {
            // When in an idle state we keep recording, but don't capture the events
            return;
        }
        var properties = {
            $snapshot_bytes: size,
            $snapshot_data: event,
            $session_id: this.sessionId,
            $window_id: this.windowId,
        };
        if (this.status !== 'disabled') {
            this._captureSnapshotBuffered(properties);
        }
        else {
            this.clearBuffer();
        }
    };
    SessionRecording.prototype._maskUrl = function (url) {
        var userSessionRecordingOptions = this.instance.config.session_recording;
        if (userSessionRecordingOptions.maskNetworkRequestFn) {
            var networkRequest = {
                url: url,
            };
            // TODO we should deprecate this and use the same function for this masking and the rrweb/network plugin
            // TODO or deprecate this and provide a new clearer name so this would be `maskURLPerformanceFn` or similar
            networkRequest = userSessionRecordingOptions.maskNetworkRequestFn(networkRequest);
            return networkRequest === null || networkRequest === void 0 ? void 0 : networkRequest.url;
        }
        return url;
    };
    SessionRecording.prototype.clearBuffer = function () {
        this.buffer = undefined;
        return {
            size: 0,
            data: [],
            sessionId: this.sessionId,
            windowId: this.windowId,
        };
    };
    // the intention is a buffer that (currently) is used only after a decide response enables session recording
    // it is called ever X seconds using the flushBufferTimer so that we don't have to wait for the buffer to fill up
    // when it is called on a timer it assumes that it can definitely flush
    // it is flushed when the session id changes or the size of the buffered data gets too great (1mb by default)
    // first change: if the recording is in buffering mode,
    //  flush buffer simply resets the timer and returns the existing flush buffer
    SessionRecording.prototype._flushBuffer = function () {
        var _this = this;
        if (this.flushBufferTimer) {
            clearTimeout(this.flushBufferTimer);
            this.flushBufferTimer = undefined;
        }
        var minimumDuration = this._minimumDuration;
        var sessionDuration = this.sessionDuration;
        // if we have old data in the buffer but the session has rotated then the
        // session duration might be negative, in that case we want to flush the buffer
        var isPositiveSessionDuration = _isNumber(sessionDuration) && sessionDuration >= 0;
        var isBelowMinimumDuration = _isNumber(minimumDuration) && isPositiveSessionDuration && sessionDuration < minimumDuration;
        if (this.status === 'buffering' || isBelowMinimumDuration) {
            this.flushBufferTimer = setTimeout(function () {
                _this._flushBuffer();
            }, RECORDING_BUFFER_TIMEOUT);
            return this.buffer || this.clearBuffer();
        }
        if (this.buffer && this.buffer.data.length !== 0) {
            this._captureSnapshot({
                $snapshot_bytes: this.buffer.size,
                $snapshot_data: this.buffer.data,
                $session_id: this.buffer.sessionId,
                $window_id: this.buffer.windowId,
            });
            return this.clearBuffer();
        }
        else {
            return this.buffer || this.clearBuffer();
        }
    };
    SessionRecording.prototype._captureSnapshotBuffered = function (properties) {
        var _this = this;
        var _a;
        var additionalBytes = 2 + (((_a = this.buffer) === null || _a === void 0 ? void 0 : _a.data.length) || 0); // 2 bytes for the array brackets and 1 byte for each comma
        if (!this.buffer ||
            this.buffer.size + properties.$snapshot_bytes + additionalBytes > RECORDING_MAX_EVENT_SIZE ||
            (!!this.buffer.sessionId && this.buffer.sessionId !== this.sessionId)) {
            this.buffer = this._flushBuffer();
        }
        if (_isNull(this.buffer.sessionId) && !_isNull(this.sessionId)) {
            // session id starts null but has now been assigned, update the buffer
            this.buffer.sessionId = this.sessionId;
            this.buffer.windowId = this.windowId;
        }
        this.buffer.size += properties.$snapshot_bytes;
        this.buffer.data.push(properties.$snapshot_data);
        if (!this.flushBufferTimer) {
            this.flushBufferTimer = setTimeout(function () {
                _this._flushBuffer();
            }, RECORDING_BUFFER_TIMEOUT);
        }
    };
    SessionRecording.prototype._captureSnapshot = function (properties) {
        // :TRICKY: Make sure we batch these requests, use a custom endpoint and don't truncate the strings.
        this.instance.capture('$snapshot', properties, {
            transport: 'XHR',
            method: 'POST',
            endpoint: this._endpoint,
            _noTruncate: true,
            _batchKey: SESSION_RECORDING_BATCH_KEY,
            _metrics: {
                rrweb_full_snapshot: properties.$snapshot_data.type === FULL_SNAPSHOT_EVENT_TYPE,
            },
        });
    };
    return SessionRecording;
}());
export { SessionRecording };
