// posthog-js/lib/src/utils/logger.js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import Config from '../config';
import { _isUndefined } from './type-utils';
import { assignableWindow, window } from './globals';
var LOGGER_PREFIX = '[PostHog.js]';
export var logger = {
    _log: function (level) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        if (window &&
            (Config.DEBUG || assignableWindow.POSTHOG_DEBUG) &&
            !_isUndefined(window.console) &&
            window.console) {
            var consoleLog = '__rrweb_original__' in window.console[level]
                ? window.console[level]['__rrweb_original__']
                : window.console[level];
            // eslint-disable-next-line no-console
            consoleLog.apply(void 0, __spreadArray([LOGGER_PREFIX], __read(args), false));
        }
    },
    info: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        logger._log.apply(logger, __spreadArray(['log'], __read(args), false));
    },
    warn: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        logger._log.apply(logger, __spreadArray(['warn'], __read(args), false));
    },
    error: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        logger._log.apply(logger, __spreadArray(['error'], __read(args), false));
    },
    critical: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        // Critical errors are always logged to the console
        // eslint-disable-next-line no-console
        console.error.apply(console, __spreadArray([LOGGER_PREFIX], __read(args), false));
    },
    uninitializedWarning: function (methodName) {
        logger.error("You must initialize PostHog before calling ".concat(methodName));
    },
};
