// @sentry/vite-plugin/dist/esm/index.mjs
import { sentryUnpluginFactory, createRollupReleaseInjectionHooks, createRollupDebugIdInjectionHooks, createRollupDebugIdUploadHooks } from '@sentry/bundler-plugin-core';
export { sentryCliBinaryExists } from '@sentry/bundler-plugin-core';

function viteReleaseInjectionPlugin(injectionCode) {
  return {
    name: "sentry-vite-release-injection-plugin",
    enforce: "pre",
    // need this so that vite runs the resolveId hook
    vite: createRollupReleaseInjectionHooks(injectionCode)
  };
}
function viteDebugIdInjectionPlugin() {
  return {
    name: "sentry-vite-debug-id-injection-plugin",
    vite: createRollupDebugIdInjectionHooks()
  };
}
function viteDebugIdUploadPlugin(upload) {
  return {
    name: "sentry-vite-debug-id-upload-plugin",
    vite: createRollupDebugIdUploadHooks(upload)
  };
}
var sentryUnplugin = sentryUnpluginFactory({
  releaseInjectionPlugin: viteReleaseInjectionPlugin,
  debugIdInjectionPlugin: viteDebugIdInjectionPlugin,
  debugIdUploadPlugin: viteDebugIdUploadPlugin
});

// eslint-disable-next-line @typescript-eslint/no-explicit-any
var sentryVitePlugin = sentryUnplugin.vite;

export { sentryVitePlugin };
